plugins/analysis-icu/src/main/java/org/elasticsearch/index/mapper/ICUCollationKeywordFieldMapper.java

```diff
@@ -0,0 +1,750 @@
+                    case "case_level":
+                        builder.caseLevel(XContentMapValues.nodeBooleanValue(fieldNode, false));
+                        iterator.remove();
+                        break;
+                    case "case_first":
+                        builder.caseFirst(XContentMapValues.nodeStringValue(fieldNode, null));
+                        iterator.remove();
+                        break;
+                    case "numeric":
+                        builder.numeric(XContentMapValues.nodeBooleanValue(fieldNode, false));
+                        iterator.remove();
+                        break;
+                    case "variable_top":
+                        builder.variableTop(XContentMapValues.nodeStringValue(fieldNode, null));
+                        iterator.remove();
+                        break;
+                    case "hiragana_quaternary_mode":
+                        builder.hiraganaQuaternaryMode(XContentMapValues.nodeBooleanValue(fieldNode, false));
+                        iterator.remove();
+                        break;
+                    default:
+                        break;
+                }
+            }
+
+            return builder;
+        }
+    }
+
+    private final String rules;
+    private final String language;
+    private final String country;
+    private final String variant;
+    private final String strength;
+    private final String decomposition;
+    private final String alternate;
+    private final boolean caseLevel;
+    private final String caseFirst;
+    private final boolean numeric;
+    private final String variableTop;
+    private final boolean hiraganaQuaternaryMode;
+    private final Collator collator;
+
+    protected ICUCollationKeywordFieldMapper(String simpleName, MappedFieldType fieldType, MappedFieldType defaultFieldType,
+                                             Settings indexSettings, MultiFields multiFields, CopyTo copyTo, String rules, String language,
+                                             String country, String variant,
+                                             String strength, String decomposition, String alternate, boolean caseLevel, String caseFirst,
+                                             boolean numeric, String variableTop, boolean hiraganaQuaternaryMode, Collator collator) {
+        super(simpleName, fieldType, defaultFieldType, indexSettings, multiFields, copyTo);
+        this.rules = rules;
+        this.language = language;
+        this.country = country;
+        this.variant = variant;
+        this.strength = strength;
+        this.decomposition = decomposition;
+        this.alternate = alternate;
+        this.caseLevel = caseLevel;
+        this.caseFirst = caseFirst;
+        this.numeric = numeric;
+        this.variableTop = variableTop;
+        this.hiraganaQuaternaryMode = hiraganaQuaternaryMode;
+        this.collator = collator.isFrozen() ? collator : collator.freeze();
```

Comment:
maybe we should just assert that the collator is frozen, there don't seem to be code paths that provide mutable collators?